### PR TITLE
Separate `web` from `http` and implement composable routers

### LINKED ISSUE #445: Make router composable
It should be possible to compose routers, so we can easily add middlewares to existing routers.

```ocaml
let router_private =
  Sihl.Web.Http.router
    ~middlewares:
      (List.concat [ middlewares; [ Middleware.Authn.middleware "/login" ] ])
    ~scope:"/"
    [ get "/ingredients" Handler.Ingredients.index
    ; post "/ingredients" Handler.Ingredients.create
    ; post "/ingredients/:name/delete" Handler.Ingredients.delete
    ; get "/pizzas" Handler.Pizzas.index
    ; post "/pizzas" Handler.Pizzas.index
    ; delete "/pizzas/:name" Handler.Pizzas.delete
    ]
;;
```

### PR DESCRIPTION
**Pull request by jerben**
Fix #445

### CODE REVIEW

**Comment by aronerben** [comment]
@jerben Is this still a draft?

File: sihl/test/web.ml [hunk 1]
```diff
@@ -1,6 +1,62 @@
 open Lwt.Syntax
```
Comment (aronerben) [inline sihl/test/web.ml:1]: Please add tests here for empty route strings, empty lists in `combine` nestings, and other edge cases

File: sihl/test/web.ml [hunk 2]
```diff
@@ -1,6 +1,62 @@
...
 open Alcotest_lwt
 
+let combine_two_routers_builds_paths _ () =
+  let router =
+    Sihl.Web.(
+      combine
+        "/root"
```
Comment (aronerben) [inline sihl/test/web.ml:8]: Let's think if we want users to have to explicitly add slashes to their routes. They might make unintended routes this way if they for example forget a slash infront of a subroute. It's also not clear what happens if you forget the slash on the root route.

File: sihl/test/web.ml [hunk 3]
```diff
@@ -1,6 +1,62 @@
...
+      (list string)
+      "builds paths"
+      [ "/root/sub/"; "/root/sub/foo"; "/root/sub/fooz"; "/root/bar" ]
+      paths);
+  Lwt.return ()
+;;
+
+let externalize_link _ () =
+  let actual = Sihl.Web.externalize_path ~prefix:"prefix" "foo/bar" in
+  Alcotest.(check @@ string) "prefixes path" "prefix/foo/bar" actual;
```
Comment (aronerben) [inline sihl/test/web.ml:52]: Please remove `@@` for all

**Review by aronerben — CHANGES_REQUESTED** [review]
Please check

File: sihl/test/web.ml (hunk 2 above)
Comment (jerben) [reply sihl/test/web.ml:8]: How about something like this:
```ocaml
let handler _ = Lwt.return @@ Sihl.Web.Response.of_plain_text ""

let router =
  Sihl.Web.(
    choose
      [ choose
          ~path:"orders"
          [ get "" handler; get ":id" handler; delete ":id" handler ]
      ; choose
          "users"
          [ get "" handler
          ; choose
              ":id"
              [ get "" handler; post ":foo" handler; get "orders" handler ]
          ]
      ])
;;
```

Comment (aronerben) [reply sihl/test/web.ml:8]: It's a little verbose, what do you think about removing the `~path` argument and making `choose` an infix operator? 
```ocaml
let router =
  Sihl.Web.(
    "/"
      >~>
      [ "orders"
          >~> [ get "" handler; get ":id" handler; delete ":id" handler ]
      ; "users"
          >~>
          [ get "" handler
          ;
              ":id"
                >~>
              [ get "" handler; post ":foo" handler; get "orders" handler ]
          ]
      ])
;;
```

File: sihl/test/web.ml (hunk 3 above)
Comment (jerben) [reply sihl/test/web.ml:52]: Fixed

File: sihl/test/web.ml (hunk 2 above)
Comment (jerben) [reply sihl/test/web.ml]: I adjusted the API now

File: sihl/test/web.ml (hunk 1 above)
Comment (jerben) [reply sihl/test/web.ml:1]: Done

**Comment by jerben** [comment]
@aronerben I reworked the API now based on our discussion. I still don't get why the infix operator is supposed to be simpler tho :smile:

File: sihl/test/web.ml [hunk 4]
```diff
@@ -1,6 +1,118 @@
...
+  Lwt.return ()
+;;
+
+let choose_routers_with_empty_scope_builds_paths _ () =
+  let router =
+    Sihl.Web.(
+      choose
+        ~scope:""
+        [ get "" (fun _ -> Lwt.return @@ Opium.Response.of_plain_text "")
+        ; get "/" (fun _ -> Lwt.return @@ Opium.Response.of_plain_text "")
```
Comment (aronerben) [inline sihl/test/web.ml:26]: It feels a little weird that you can add a slash with this. Shouldn't you have to create another subroute here which is also `""` to get that slash?

File: sihl/test/web.ml (hunk 1 above)
Comment (aronerben) [reply sihl/test/web.ml:1]: Thanks

**Comment by aronerben** [comment]
> @aronerben I reworked the API now based on our discussion. I still don't get why the infix operator is supposed to be simpler tho

Hehe, my idea was because the pattern felt like `cons` vs `::`, so something like 
```ocaml
choose "a" (choose "b" (choose "c" [])) 
"a" >-> "b" >-> "c" >-> []
```
But since the second argument for the operator is a list anyway it doesn't add too much, so I'm fine with leaving it as a prefix operator.

**Review by aronerben — APPROVED** [review]

**Review by aronerben — COMMENTED** [review]
Please check

File: sihl/test/web.ml (hunk 4 above)
Comment (jerben) [reply sihl/test/web.ml:26]: You should if you want to use the routers correctly. This is a demonstration of incorrect use where we try to do the right thing.
I am thinking about a `sihl-doctor` package that has a command `sihl doctor` that lists all the things you might be doing wrong. I think this is nicer than `Log.info` or even exceptions.

Comment (aronerben) [reply sihl/test/web.ml:26]: Good idea, let's do it like that.
- [ ] Add docs for `get`
- [ ] Add a comment that `sihl-doctor` checks this

Comment (jerben) [reply sihl/test/web.ml:26]: @aronerben I just realized that we still have to allow using slashes in case someone doesn't want to use routers and just wants to create a URL `/foo/bar/my/long/path` for some reason. (Maybe a monitoring tool polling a hard coded URL or something like that.

Comment (jerben) [reply sihl/test/web.ml:26]: I created an issue for `sihl doctor` in general though

**Comment by aronerben** [comment]
@jerben Approved, can be merged when you're ready

**Comment by jerben** [comment]
Thanks @aronerben
